File: String_to_Integer/python/String_to_integer.py
class Solution:
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        str = str.strip()
        if not str: return 0

        ans = 0
        sign = 1
        INT_MAX = 2 ** 31 -1 # 2147483647
        INT_MIN = - 2 ** 31

        if(str[0] == '+' or str[0] == '-'):
            sign = 1 if str[0] == '+' else -1
            str = str[1:]
        idx = 0

        while(idx < len(str) and str[idx] >= '0' and str[idx] <= '9'):

            if(ans > INT_MAX/10 or (ans == 214748364 and str[idx] > '7')):
                return INT_MAX if sign == 1 else INT_MIN
            ans = ans*10 + ord(str[idx]) - ord('0')
            idx += 1

        return sign*ans

if __name__ == "__main__":
  str = "4193 with words"
  ans = Solution()

  print("Input: {}".format(str))
  print("Output: {}".format(ans.myAtoi(str)))

File: problems/stringToInteger/c++/String_to_integer.cpp
//
//  String_to_integer.cpp
//  
//
//  Created by Wen-Ting Wang on 2018/11/28.
//
#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
    int myAtoi(string str) {
        
        if(str.empty()) return 0;
        
        int ans = 0, len = str.size(), i = 0, sign = 1;
        
        while(i < len && str[i] == ' ') ++i;
        
        if(str[i] == '+' || str[i] == '-')
            sign = (str[i++] == '+') ? 1 : -1;
        
        while(str[i] >= '0' && str[i] <= '9' && i < len){
            
            if(ans  > INT_MAX / 10 || ( ans  == INT_MAX / 10 && str[i] - '0' > 7))
                return (sign == 1) ? INT_MAX : INT_MIN;
            
            ans = 10 * ans + (str[i++] - '0');
        }
        
        return ans * sign;
    }
};

int main(){
    string str = "4193 with words";
    
    Solution ans;
    
    cout << "Input: " << str << endl;
    cout << "Output: "<< ans.myAtoi(str) << endl;

}
